W-15082979-dw-expression-builder-advanced-config-duke (#296)
mulesoft/docs-code-builder: modules/ROOT/pages/int-configure-components.adoc

= Configuring Components
:open-config-xml: In the Explorer, open the configuration XML file for your project:

Configure connectors and processors within your implementation and integration projects from the canvas or configuration XML. 

Common tasks include the following: 

* <<add-components>>
* xref:int-work-with-code-snippets.adoc[Work with Code Snippets]
* xref:int-configure-dw-expressions.adoc[Configure Expressions]
* <<check-connection>>
* <<import-connectors-from-exchange>>
* <<open-component-from-xml>>

== Before You Begin

* Determine which connectors, processors, and snippets you intend to use.
+
// Note on ways to start component configs: 
include::partial$acb-component-info.adoc[tags="note-component-add-config"]
// Pointer to component reference:
include::partial$acb-component-info.adoc[tags="note-component-ref"]

* Review xref:index.adoc#get-familiar-with-the-basics[Get Familiar with the Basics] to ensure that you are familiar with important concepts and features. 
+

> Understand the basics of Mule flows, the Mule event structure, including the payload, attributes, and Mule variables. To add transform data and add expressions to your components, have some familiarity of the DataWeave language.

Author: ```suggestion
Understand the basics of Mule flows, the Mule event structure, including the payload, attributes, and Mule variables. To transform data and add expressions to your components, have some familiarity of the DataWeave language.
```

[[add-components]]
== Add a Component to Your Project

> Add components to your project. The following example illustrates basic configurations from the canvas UI and configuration XML. The example assumes you are beginning with an empty integration project.

Reviewer: ```suggestion
The following example illustrates basic configurations for adding components to your project from the canvas UI, and the configuration XML. The example assumes you are beginning with an empty integration project.
```

. {open-config-xml}, such as `my-project-name.xml`. 
+
image::int-empty-canvas.png["Canvas showing visual representation of Mule flow and the Mule configuration file"]
+
[calloutlist]
.. The canvas provides space for a visual representation of your Mule flows or subflows.
.. The configuration XML editor displays the configuration file for your Mule application.
. Select *Build a Flow* from the canvas to create an empty flow within a Mule integration application.
. Change the default name of the flow from the canvas UI or from the configuration XML.
+
[tabs]
====
From the canvas UI::
+
--
Click *Flow name1* to open the configuration panel for the Flow component, change the flow name, and click the check mark to set the new name. 

image::int-flow-name-ui.png["Change name of flow through canvas UI."]
--
From the configuration XML::
+
--

> Replace the default name of the flow (`name1`) with your flow name, such as `getFlights`. For example:

Reviewer: ```suggestion
Replace the default name of the flow (`name1`) with your flow name, such as `getFlights`, for example:
```

Reviewer: Kim's guidance, not mind 🤣

Author: so hilarious, Diane and Beth both suggested "For example: " 
I always used ", for example: " before that. ¯\_(ツ)_/¯

Reviewer: https://confluence.internal.salesforce.com/display/MTDT/MuleSoft+CX+Writing+Style+Reference#MuleSoftCXWritingStyleReference-%22forexample%22

[source,xml]
----
<flow name="my-flow" >
    
</flow>
--
====

> . Add a component to your project from the canvas, for example, add the HTTP Listener component:

Reviewer: See?!

> . Add a component to your project from the canvas, for example, add the HTTP Listener component:

Reviewer: ```suggestion
. Add a component to your project from the canvas.
+
For example, add the HTTP Listener component:
```

Reviewer: (To make it more clear that the sublist is an example.)

.. In the canvas, click the image:icon-plus.png["",15,15] (*Add component*) icon.
.. In the *Add Component* panel, search for and select *Listener* from the *HTTP* results:
+
image::main-tutorial-add-first-listener.png["Listener component highlighted in the Add Component section"]
+

> The configuration XML file now includes the XML for the HTTP Listener into the `<flow/>` element, for example:

Reviewer: ```suggestion
The configuration XML file now includes the XML for the HTTP Listener within the `<flow/>` element, for example:
```

Reviewer: ("into" feels weird here)

+
[source,XML]
----
<flow name="getFlights" >
  <http:listener path="mypath" config-ref="config-ref" doc:name="Listener" doc:id="rrjiqa" />

</flow>
----
. Add another component, this time using the XML configuration menu.

> For example, add the `<http:listener-config/>` component from a snippet.

Reviewer: ```suggestion
+
For example, add the `<http:listener-config/>` component from a snippet.
```

+
For more information about snippets, see xref:int-work-with-code-snippets.adoc[].
+
.. In the configuration XML, place your cursor before the opening `<flow>` tag, and type `http`.
+
Ensure that the cursor is not inside the `<flow/>` element.
+
.. Type `http`, and select the `http:listener-config` snippet:
+
image::add-http-config-snippet.png["http:listener-config snippet highlighted in the configuration XML menu"]
+
The snippet adds the following code:
+
[source,xml]
----
<http:listener-config name="HTTP_Listener_config" >
  <http:listener-connection host="0.0.0.0" port="8081" />
</http:listener-config>
----

> .. Notice that the Listener component in the canvas now displays an error:

Reviewer: Is this a step? Or should it be a hanging paragraph?

Author: nope. good catch

Author: actually, yes, it's a step. I need them to see the error. Could reword  ¯\_(ツ)_/¯

Reviewer: In general, I think of "notice" as not a user action.

+
image::int-canvas-error.png["Listener error in the canvas"]
+

> To determine where the error is, select the processor in the canvas.

Reviewer: ```suggestion
.. To determine where the error is, select the processor in the canvas.
+
```

> Anypoint Code Builder highlights its location within the configuration XML, and you can mouse over the issue for more information. For example:

Reviewer: ```suggestion
Anypoint Code Builder highlights its location within the configuration XML, and you can mouse over the issue for more information, for example:
```

* To add a DataWeave expression or a Mule variable as a value, such as `payload`, click *fx* (located before the field), and provide the value, for example:
+
image::int-set-payload-config-fx.png["Adding expression to Set Payload"]
+
For more information about configuring DataWeave expressions, see xref:int-configure-dw-expressions.adoc[].

* To add a configuration property as a value, type a value such as `${secure::mysensitiveprop}`. For example:
+
image::int-set-payload-config-property.png["Adding configuration property to Set Payload"]
+
For more information about configuration properties, see xref:int-create-secure-configs.adoc[].

Your configuration XML file now looks similar to the following:

[source,xml]
--
<http:listener-config name="config-ref" >
  <http:listener-connection host="0.0.0.0" port="8081" />
</http:listener-config>

<flow name="getFlights" >
  <http:listener path="path" config-ref="config-ref" doc:name="Listener" doc:id="rrjiqa" />
  <set-payload value="my value" doc:name="Set payload" doc:id="gecykt" />
    
</flow>
--

[[import-connectors-from-exchange]]
== Import a Connector from Exchange

> Anypoint Connectors provide operations for retrieving, modifying, and sending data to and from systems.

Reviewer: You could switch this around to be user-value first (To retrieve, .... use...)

Author: todo to self: review

Author: want to emphasize what connectors do in this case, as a description vs. an action

In addition to the xref:acb-reference.adoc#builtin-connectors[built-in connectors] that Anypoint Code Builder provides,
you can download many other connectors from Anypoint Exchange.

* <<import-canvas>>
* <<import-command>>

[[import-canvas]]
=== Import a Connector from the Canvas

To import a connector from Exchange and add it to your configuration:

//Open config XML:
. {open-config-xml}, such as `my-project-name.xml`.
. Click image:icon-tree-view.png["",15,15] (*Show Mule graphical mode*) in the activity bar to open the UI canvas if it doesn't open automatically.
. Add the connector the same way you added other components from the canvas:
.. In the canvas, click the image:icon-plus.png["",15,15] (*Add component*) icon.
.. In the *Add Component* panel, click *Connectors*.
.. Click the connector name and then click the operation to add, such as *Publish*:
+
image::int-add-connector-operation.png["Add Publish operation from the Anypoint MQ Connector"]
+

> If the connector is not available locally, click the image:icon-search-exchange.png["",15,15] (*Search in Exchange*) toggle:

Reviewer: step?

Author: todo to self: need to build and check

Author: converted

//Open config XML:
. {open-config-xml}, such as `my-project-name.xml`.
. Click image:icon-tree-view.png["",15,15] (*Show Mule graphical mode*) in the activity bar to open the canvas UI if it doesn't open automatically.
// Pointer to Command Palette
include::partial$acb-reusable-steps.adoc[tags="open-command-palette"]
. Select the following command:
+
[source,command]
----
MuleSoft: Import Asset from Exchange
----
. Select *Connector*.
. Search for the connector name to import, such as "MQ," for example:
+
image::int-import-connect-exchange.png["Import Anypoint MQ Connector from Exchange]
. Select the connector.
. At the prompt, select the version of the connector to import, such as `v4.0.3`.
+
Anypoint Code Builder imports the connector and makes it available in the Components list.

For more information about the connectors available on Exchange, see 
xref:connectors::index.adoc[]. 
For more information about Exchange, see 
https://anypoint.mulesoft.com/exchange/[Exchange (US)^] or
https://eu1.anypoint.mulesoft.com/exchange/[Exchange (EU)^]
and
xref:exchange::index.adoc[].

[[check-connection]]
== Test a Connection Configuration

> Most connectors provide an operation that triggers the flow to start, such as the HTTP Listener operation. These operations typically have connection configurations that you can test from the configuration XML. For example, to test an HTTP Listener connection, click *Test Connection* in the configuration XML:

Reviewer: ```suggestion
Most connectors provide an operation, such as the HTTP Listener operation, that triggers the flow to start. These operations typically have connection configurations that you can test from the configuration XML. For example, to test an HTTP Listener connection, click *Test Connection* in the configuration XML:
```

image::http-listener-test-connection.png["HTTP Listener Test Connection link in the configuration XML"]

The status bar shows the progress:

* *Verifying connection inbound-request* indicates that the test is in progress.
* *Connection is valid* indicates a successful connection. 

> * *Invalid Connection Got status code: 500 when trying to resolve a Mule Runtime operation* indicates a connection error.

Reviewer: Can we copyedit the errors? 🤣

Author: unfortunately, we can't edit some of them, such as this one.

+
A common code `500` error is `port 8081: Address already in use`. For port conflicts, configure a different port, such as `8082`, and retest the connection. 

Before publishing, run your application in debug mode.

[[open-component-from-xml]]
== Open a Component in the Canvas from the XML Editor

> From the configuration XML, open a component and its configuration panel in the canvas. This feature is helpful, for example, if you want to configure a component from the UI, if a component is in another flow than the flow currently displayed in the canvas, or if the canvas is not open.  

Reviewer: You could switch this around and say "If you want to configure, ..., ..., open a component..."

Author: todo to self: review

Author: reworded a bit

    
</flow>
--

[[import-connectors-from-exchange]]
== Import a Connector from Exchange

Anypoint Connectors provide operations for retrieving, modifying, and sending data to and from systems.
In addition to the xref:acb-reference.adoc#builtin-connectors[built-in connectors] that Anypoint Code Builder provides,
you can download many other connectors from Anypoint Exchange.

* <<import-canvas>>
* <<import-command>>

[[import-canvas]]
=== Import a Connector from the Canvas

To import a connector from Exchange and add it to your configuration:

//Open config XML:
. {open-config-xml}, such as `my-project-name.xml`.
. Click image:icon-tree-view.png["",15,15] (*Show Mule graphical mode*) in the activity bar to open the UI canvas if it doesn't open automatically.
. Add the connector the same way you added other components from the canvas:
.. In the canvas, click the image:icon-plus.png["",15,15] (*Add component*) icon.
.. In the *Add Component* panel, click *Connectors*.
.. Click the connector name and then click the operation to add, such as *Publish*:
+
image::int-add-connector-operation.png["Add Publish operation from the Anypoint MQ Connector"]

> +

Author: ```suggestion
```

> If the connector is not available locally, click the image:icon-search-exchange.png["",15,15] (*Search in Exchange*) toggle:

Author: ```suggestion
.. If the connector is not available locally, click the image:icon-search-exchange.png["",15,15] (*Search in Exchange*) toggle:
```

https://eu1.anypoint.mulesoft.com/exchange/[Exchange (EU)^]
and
xref:exchange::index.adoc[].

[[check-connection]]
== Test a Connection Configuration

Most connectors provide an operation, such as the HTTP Listener operation, that triggers the flow to start. These operations typically have connection configurations that you can test from the configuration XML. For example, to test an HTTP Listener connection, click *Test Connection* in the configuration XML:

image::http-listener-test-connection.png["HTTP Listener Test Connection link in the configuration XML"]

The status bar shows the progress:

* *Verifying connection inbound-request* indicates that the test is in progress.
* *Connection is valid* indicates a successful connection. 
* *Invalid Connection Got status code: 500 when trying to resolve a Mule Runtime operation* indicates a connection error.
+
A common code `500` error is `port 8081: Address already in use`. For port conflicts, configure a different port, such as `8082`, and retest the connection. 

Before publishing, run your application in debug mode.

[[open-component-from-xml]]
== Open a Component in the Canvas from the XML Editor

> From the configuration XML, open a component and its configuration panel in the canvas. This feature is helpful, for example, if you want to configure a component from the UI, if a component is in another flow than the flow currently displayed in the canvas, or if the canvas is not open.  

Author: ```suggestion
You can open a component and its configuration panel in the canvas from the XML editor. This feature is helpful, for example, if you want to configure the component from the UI, if a component is in another flow than the flow currently displayed in the canvas, or if the canvas is not open.  
```
